Source/NSPathCell.m

```diff
@@ -0,0 +1,359 @@
+}
+
+- (void) setPathComponentCells: (NSArray *)cells
+{
+  ASSIGNCOPY(_pathComponentCells, cells);
+}
+
+- (SEL) doubleAction
+{
+  return _doubleAction;
+}
+
+- (void) setDoubleAction: (SEL)action
+{
+  _doubleAction = action;
+}
+
+- (id) objectValue
+{
+  return _objectValue;
+}
+
+- (void) setObjectValue: (id)obj
+{
+  ASSIGN(_objectValue, obj);
+  [self setPathComponentCells:
+          [NSPathCell _generateCellsForURL: (NSURL *)_objectValue]];
+}
+
+- (NSURL *) URL
+{
+  return [self objectValue];
+}
+
+- (void) setURL: (NSURL *)url
+{
+  [self setObjectValue: url];
+}
+
+- (id<NSPathCellDelegate>) delegate
+{
+  return _delegate;
+}
+
+- (void) setDelegate: (id<NSPathCellDelegate>)delegate
+{
+  _delegate = delegate;
+}
+
+- (void) drawInteriorWithFrame: (NSRect)frame inView: (NSView *)controlView
+{
+  NSUInteger count = [_pathComponentCells count];
+
+  [super drawInteriorWithFrame: frame
+                        inView: controlView];
+  if (count > 0)
+    {
+      NSEnumerator *en = [_pathComponentCells objectEnumerator];
+      NSPathComponentCell *cell = nil;
+      
+      while ((cell = (NSPathComponentCell *)[en nextObject]) != nil)
+        {
+          NSRect f = [self rectOfPathComponentCell: cell
+                                         withFrame: frame
+                                            inView: controlView]; 
+          [cell drawInteriorWithFrame: f
+                               inView: controlView];
+        }
+    }
+}
+
+- (id) initWithCoder: (NSCoder *)coder
+{
+  self = [super initWithCoder: coder];
+  if ([coder allowsKeyedCoding])
+    {
+      [self setPathStyle: NSPathStyleStandard];
+      
+      if ([coder containsValueForKey: @"NSPathStyle"]) 
+        {
+          [self setPathStyle: [coder decodeIntegerForKey: @"NSPathStyle"]];
+        }
+    }
+  else
+    {
+      [coder decodeValueOfObjCType: @encode(NSUInteger)
+                                at: &_pathStyle];
+      [self setPathComponentCells: [coder decodeObject]];
+    }
+
+  return self;
+}
+
+- (void) encodeWithCoder: (NSCoder *)coder
+{
+  [super encodeWithCoder: coder];
+  if ([coder allowsKeyedCoding])
+    {
+      [coder encodeInteger: [self pathStyle]
+                    forKey: @"NSPathStyle"];
+      [coder encodeObject: [self pathComponentCells]
+                   forKey: @"NSPathComponentCells"];
+    }
+  else
+    {
+      [coder encodeValueOfObjCType: @encode(NSUInteger)
+                                at: &_pathStyle];
+      [coder encodeObject: [self pathComponentCells]];
+    }
+}
+
+@end
+
+@implementation NSPathCell (Private)
+
+// Private...
++ (NSArray *) _generateCellsForURL: (NSURL *)url
+{
+  NSMutableArray *array = [NSMutableArray arrayWithCapacity: 10];
+  
+  // Create cells
+  if (url != nil)
+    {
+      BOOL isDir = NO;
+      BOOL at_root = NO;
+      NSFileManager *fm = [NSFileManager defaultManager];
+      NSURL *u = [url copy];
+      
+      // Decompose string...
+      do
```

Comment:
I prefer while loops and in this specific case it would make no difference to move the condition up as at_root is NO at the start.